packages/loader/container-loader/src/container.ts

```diff
@@ -1629,7 +1629,7 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         }
 
         const state = this.connectionState === ConnectionState.Connected;
-        if (!this.context.disposed) {
+        if (this._context?.disposed === false) {
```

Comment:
This line itself looks fine, but I do not see how we can get here on container closure. Any chance you can clarify / provide stack?
I just wonder if the fix should be in another place, given that we should get into propagateConnectionState() only when container loaded.
Maybe loaded() getter is wrong? I.e. if we switched from "loading" to "closing" / "closed", then I can see how we get here, but I'd think that the fix then should be to adjust loaded (I actually think that logic around this._lifecycleState became too complicated and too hidden potentially causing these issues)

Reply:
i hit it when testing https://dev.azure.com/fluidframework/internal/_workitems/edit/605

we also see them all the time in automation: 

union office_fluid_ffautomation_error
| where Data_eventName == "fluid:telemetry:Container:ContainerCloseException"
| project-reorder Event_Time, Data_branch, Data_testName, Data_testVariant, Data_eventName, Data_error, Data_stack

Reply:
generally, i think the getter that throws is a bad pattern and lead to wrong assumptions. happy to close this, and let someone else take on [AB#625](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/625)

Reply:
I've say - AB#625 should be handled independently from this bug.
Looking at the code, the only way to have _context === undefined if we did not finish load.
And based on the callstack below, it looks like close() is called from static load() function due to error on loading path.
I think the core of the issues here is that this.loaded becomes true once we enter Container.close().

I believe that's regression from https://github.com/microsoft/FluidFramework/pull/7933.
Reason why I do not want to fix it that way - other things that rely on this.loaded (in the future) can be broken in similar way, so I'd rather ensure it has proper semantics, and deal with other issues independently. Or maybe it should return false once we move to closing/closed state (not sure, but I'd prefer more predictable earlier behavior).
The other alternative - remove loaded() getter and adjust condition in connectionStateChanged handler to only propagate events if we are loaded and not closed. (i.e. force users to consider all actual states and not provide them helper getters like loaded() that may cause misunderstanding).

The callstack when it happens looks like that:

    at Container.get context [as context] (/mnt/vss/_work/1/test/node_modules/@fluidframework/test-version-utils/node_modules/.legacy/0.58.2002/node_modules/@fluidframework/container-loader/dist/container.js:356:19)
    at Container.propagateConnectionState (/mnt/vss/_work/1/test/node_modules/@fluidframework/test-version-utils/node_modules/.legacy/0.58.2002/node_modules/@fluidframework/container-loader/dist/container.js:1110:19)
    at Object.connectionStateChanged (/mnt/vss/_work/1/test/node_modules/@fluidframework/test-version-utils/node_modules/.legacy/0.58.2002/node_modules/@fluidframework/container-loader/dist/container.js:193:26)
    at ConnectionStateHandler.setConnectionState (/mnt/vss/_work/1/test/node_modules/@fluidframework/test-version-utils/node_modules/.legacy/0.58.2002/node_modules/@fluidframework/container-loader/dist/connectionStateHandler.js:206:22)
    at ConnectionStateHandler.receivedDisconnectEvent (/mnt/vss/_work/1/test/node_modules/@fluidframework/test-version-utils/node_modules/.legacy/0.58.2002/node_modules/@fluidframework/container-loader/dist/connectionStateHandler.js:129:14)
    at DeltaManager.<anonymous> (/mnt/vss/_work/1/test/node_modules/@fluidframework/test-version-utils/node_modules/.legacy/0.58.2002/node_modules/@fluidframework/container-loader/dist/container.js:1035:41)
    at DeltaManager.emit (events.js:400:28)
    at DeltaManager.disconnectHandler (/mnt/vss/_work/1/test/node_modules/@fluidframework/test-version-utils/node_modules/.legacy/0.58.2002/node_modules/@fluidframework/container-loader/dist/deltaManager.js:384:14)
    at Object.disconnectHandler (/mnt/vss/_work/1/test/node_modules/@fluidframework/test-version-utils/node_modules/.legacy/0.58.2002/node_modules/@fluidframework/container-loader/dist/deltaManager.js:59:49)
    at ConnectionManager.disconnectFromDeltaStream (/mnt/vss/_work/1/test/node_modules/@fluidframework/test-version-utils/node_modules/.legacy/0.58.2002/node_modules/@fluidframework/container-loader/dist/connectionManager.js:448:20)
    at ConnectionManager.dispose (/mnt/vss/_work/1/test/node_modules/@fluidframework/test-version-utils/node_modules/.legacy/0.58.2002/node_modules/@fluidframework/container-loader/dist/connectionManager.js:249:14)
    at DeltaManager.close (/mnt/vss/_work/1/test/node_modules/@fluidframework/test-version-utils/node_modules/.legacy/0.58.2002/node_modules/@fluidframework/container-loader/dist/deltaManager.js:360:32)
    at Container.close (/mnt/vss/_work/1/test/node_modules/@fluidframework/test-version-utils/node_modules/.legacy/0.58.2002/node_modules/@fluidframework/container-loader/dist/container.js:465:36)
    at /mnt/vss/_work/1/test/node_modules/@fluidframework/test-version-utils/node_modules/.legacy/0.58.2002/node_modules/@fluidframework/container-loader/dist/container.js:301:27

Reply:
The other way to say it - propagateConnectionState() should assert that this._context !=== undefined :) Code should enforce this invariant.

Reply:
i'm unopinionated on the fix, but this is causing a lots of noise in our test runs, so i'd like to see if fixed sooner than later. if #7993  was the culprit, maybe @markfields  can take a look at fixing it

Reply:
I just skimmed this but didn't look closely at the code like you guys did.  If there's a follow-up bug I or someone should dig into more, please open it in ADO.